transports/quic/src/endpoint.rs

```diff
@@ -402,18 +402,41 @@ where
 
         if me.event_slot.is_none() && me.incoming_slot.is_none() {
             let mut metas = [RecvMeta::default(); BATCH_SIZE];
-            let mut iovs = MaybeUninit::<[IoSliceMut; BATCH_SIZE]>::uninit();
-            me.recv_buf
-                .chunks_mut(me.recv_buf.len() / BATCH_SIZE)
-                .enumerate()
-                .for_each(|(i, buf)| unsafe {
-                    iovs.as_mut_ptr()
-                        .cast::<IoSliceMut>()
-                        .add(i)
-                        .write(IoSliceMut::new(buf));
-                });
-            let mut iovs = unsafe { iovs.assume_init() };
-            while let Poll::Ready(result) = me.socket.poll_recv(cx, &mut iovs, &mut metas) {
+            let mut iovs = unsafe {
+                // TODO: use MaybeUninit::uninit_array when stabilized
+                // SAFETY: An uninitialized `[MaybeUninit<_>; LEN]` is valid.
+                MaybeUninit::<[MaybeUninit<IoSliceMut>; BATCH_SIZE]>::uninit().assume_init()
+            };
+            // SAFETY: safe as long as iovs do not outlive recv_buf
+            // and there exists only one iovs initialized with the given recv_buf.
+            unsafe fn init_iovs<'a, 'b>(
+                iovs: &'b mut [MaybeUninit<IoSliceMut>],
+                recv_buf: &'a mut [u8],
+            ) -> &'b mut [IoSliceMut<'b>] {
+                // every iovs elem must be initialized with an according elem from buf chunks
+                debug_assert_eq!(
+                    recv_buf.chunks_mut(recv_buf.len() / BATCH_SIZE).len(),
+                    iovs.len()
+                );
+                recv_buf
+                    .chunks_mut(recv_buf.len() / BATCH_SIZE)
+                    .enumerate()
+                    .for_each(|(i, buf)| unsafe {
+                        iovs.as_mut_ptr()
+                            .cast::<IoSliceMut>()
+                            .add(i)
+                            .write(IoSliceMut::new(buf));
+                    });
+
+                // TODO: use MaybeUninit::slice_assume_init_mut when stabilized
+                // SAFETY: casting slice to a `*mut [T]` is safe since the caller guarantees that
+                // `slice` is initialized, and `MaybeUninit` is guaranteed to have the same layout as `T`.
+                // The pointer obtained is valid since it refers to memory owned by `slice` which is a
+                // mutable reference and thus guaranteed to be valid for writes.
+                &mut *(iovs as *mut [MaybeUninit<IoSliceMut>] as *mut [IoSliceMut])
```

Comment:
is this really more straightforward than the previous solution?

Reply:
It's not, however it will be replaces with slice_assume_init_mut (not a big deal compared to assume_init). The issue is to work with reference, not with a stack allocated obj of 512 bytes len. I can revert it if you like.

Reply:
What's your decision?

Reply:
MaybeUninit::assume_init_mut is in stable so I've rewritten the code and simplified it a bit. Fewer unsafes than before.